CI: workflow for building weels

fabinsch opened the pull request:
- add package.xml, pyproject.toml, setup.cfg
- new workflow for release

fabinsch committed:
CI: workflow for building weels

- add package.xml, pyproject.toml, setup.cfg
- new workflow for release

fabinsch commented:
this here is almost done, except for the last part in the `release.yml` which is uploading the three wheels with twine.

during the creating of this pipeline I was using the CI, here you can see the latest test that passed: https://github.com/jcarpent/proxqp-unofficial/actions/runs/2804412577 .

jcarpent commented on package.xml:
```diff
@@ -0,0 +1,23 @@
+  <build_depend>doxygen</build_depend>
+  <depend>eigen</depend>
+  <depend>simde</depend>
+  <depend>pybind11</depend>
```
You can remove this.

jcarpent commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+[build-system]
+requires = ["cmeel[build]"]
+build-backend = "cmeel.build"
+configure-args = ["-DCMAKE_CXX_STANDARD=17","-DCMAKE_CXX_COMPILER=gcc","-DBUILD_TESTING=OFF","-DBUILD_PYTHON_INTERFACE=ON","-DBUILD_WITH_VECTORIZATION_SUPPORT=ON","-DINSTALL_DOCUMENTATION=OFF"]
```
```suggestion
configure-args = ["-DCMAKE_CXX_COMPILER=gcc","-DBUILD_TESTING:BOOL=OFF","-DBUILD_PYTHON_INTERFACE:BOOL=ON","-DBUILD_WITH_VECTORIZATION_SUPPORT:BOOL=ON","-DINSTALL_DOCUMENTATION:BOOL=OFF"]
```

jcarpent replied:
Could we use clang instead of gcc? Clang generates more efficient binaries.

fabinsch replied:
I tried it before and it did not work. I will try it again, for this I will trigger a pipeline here and see if it works with clang and without defining the c++ standard.

fabinsch replied:
after removing the standard here i get the following error in my pipeline:

```
      -- Default C++ standard:
      -- Minimal C++ standard upgraded to 17
      CMake Error at cmake-module/cxx-standard.cmake:101 (message):
        Unknown current C++ standard while trying to check for >= 17
```

(https://github.com/jcarpent/proxqp-unofficial/runs/7695050018?check_suite_focus=true#step:7:87)

why you think it is not ok to input it here, actually I can set it to whatever I want just it needs to be initialized I think.

jcarpent replied:
I think this is because the cmake module is not up-to-date. Have you rebased on latest devel branch?

fabinsch replied:
you are right. after checking out manually the latest master of the cmake-module it was fine. we should update this in the repo. now I have another problem with clang that I saw before and that made me use gcc. I think we have somehow problems with the standard library bc the manylinux docker that we are using are supposed to use the GCC (I think..)

```
      In file included from /__w/proxqp-unofficial/proxqp-unofficial/bindings/python/external/pybind11/include/pybind11/stl.h:22:
      /__w/proxqp-unofficial/proxqp-unofficial/3/envs/proxsuite/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/include/c++/10.4.0/valarray:1249:5: error: exception specification in declaration does not match previous declaration
          end(const valarray<_Tp>& __va) noexcept
          ^
      /__w/proxqp-unofficial/proxqp-unofficial/3/envs/proxsuite/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/include/c++/10.4.0/bits/range_access.h:110:37: note: previous declaration is here
        template<typename _Tp> const _Tp* end(const valarray<_Tp>&);
```

https://github.com/jcarpent/proxqp-unofficial/runs/7695886810?check_suite_focus=true#step:8:2439

jcarpent replied:
OK. Keep then gcc. we will fix it later then.

fabinsch replied:
ok, I switched back to gcc, pipeline is passing.

jcarpent commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,70 @@
+name: Release on GitHub & PyPI
+
+on:
+  push:
+    tags:
+      - "v*"
```
On release, no?

fabinsch replied:
i think it is fine like this. from my point of view, we want to start this pipeline here by just doing a tag with `vX.X.X`, which then builds the wheel and does the release, check the end of the workflow. 
When this workflow creates the release, it will trigger the gh-pages workflow which is on release. if you have another concept in mind let me know.

jcarpent replied:
I think we need two steps, when merging a PR, we should check that the wheel creation works.
When releasing a new version, we should publish it.

jcarpent replied:
What do you think?

fabinsch replied:
ahhh, okay. I see your point. so the first part of this pipeline should run all the time ? or just for merging a PR?
And then the lower part should just run if we put a tag, to signal we want to release. this triggers the upper part, plus the lower part that actually creates the github release and uploads the wheels to pypip.

jcarpent replied:
Maybe all the time, but it is quite expansive.
Yes for the second part.

fabinsch replied:
mmm, I am wondering... at the end building the wheel using cmeel (https://github.com/cmake-wheel/cmeel) boils down to using cmake with the same setup as in our ci-linux-osx-windows. so maybe it is enough to run it for a PR to make sure, if cmeel changes etc but I don't see how a commit could break the wheels but not the other pipeline.

jcarpent replied:
Fine for me.

fabinsch replied:
Ok, I did the changes in this way now. Pipeline first part (building wheels) should run on every PR + release, second part (uploading) should just be for release in our repo (done the same way as  https://github.com/robotology/idyntree/blob/master/.github/workflows/python.yml)

fabinsch committed:
ci/release: use remove c++ standard

- temp. remove the on branch -v to see a pipeline running

fabinsch committed:
ci/wheels: run building wheel on each PR, publish pypip for release

fabinsch committed:
update cmake submodule

fabinsch committed:
update cmake submodule in workflow

fabinsch committed:
ci: publish wheels on pypi test

fabinsch committed:
ci: build wheels for macos

fabinsch committed:
readme: fix display logo on pypi

- need to change from relative path to absolute path to not have a broken image on pypi
- it will be broken, until the simple-robotics/proxsuite is public

fabinsch committed:
readme: mention installation via pip

fabinsch commented:
@jcarpent In my opinion this here is done. Wheels could be uploaded to pypi test and work on Ubuntu20 and MacOs. I edit slightly the readme, check if this is fine for you.

jcarpent commented:
Could you rebase on the current devel?

fabinsch committed:
pip: publish v0.0.0

- have a first version for linux and macos on pypi
- if condition needs to be turned on again after successful first upload

fabinsch committed:
ci/wheel: upload only proxsuite wheels
